packages/credentials/src/V1/KiltCredentialV1.ts

```diff
@@ -327,15 +327,77 @@ export function fromInput({
 
 const cachingCTypeLoader = newCachingCTypeLoader()
 
+// Check recursively if a value has references
+const hasRef = (value: unknown): boolean => {
+  if (typeof value !== 'object' || value === null) {
+    return false
+  }
+
+  if ('$ref' in (value as Record<string, unknown>)) {
+    return true
+  }
+
+  if (Array.isArray(value)) {
+    return value.some((item) => hasRef(item))
+  }
+
+  return Object.values(value as Record<string, unknown>).some((v) => hasRef(v))
+}
+
+// Single function to both check for references and extract them
+function extractUniqueReferences(
+  cType: ICType,
+  references: Set<string> = new Set<string>()
+): Set<string> {
+  if (typeof cType?.properties !== 'object' || cType.properties === null) {
+    return references
+  }
+
+  const processValue = (value: unknown): void => {
+    if (typeof value !== 'object' || value === null) {
+      return
+    }
+    const objValue = value as Record<string, unknown>
+    if ('$ref' in objValue) {
+      const ref = objValue.$ref as string
+      if (ref.startsWith('kilt:ctype:')) {
+        references.add(ref.split('#/')[0])
+      }
+    }
+
+    if (Array.isArray(value)) {
+      value.forEach(processValue)
+    } else {
+      Object.values(objValue).forEach(processValue)
```

Comment:
the point of passing down the set of references to extractUniqueReferences was that you can call extractUniqueReferences in recursion instead of an inner function. You don't necessarily need to do it that way, but if you're not using this parameter, you can also just pull it back into the function as a normal variable.

Reply:
Fixed at: https://github.com/KILTprotocol/sdk-js/pull/919/commits/bbfced24e1897748e57f40ddc5fa06267ed4962c